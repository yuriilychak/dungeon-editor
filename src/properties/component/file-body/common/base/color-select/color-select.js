import React, {memo, useCallback} from "react";
import ColorPicker from "rc-color-picker";
import {string, func, bool} from "prop-types";

import "rc-color-picker/assets/index.css";
import "./color-select.scss";
import {FIELD_TYPE} from "../../../../../constants";
import {generateChangeEvent} from "../../helpers";

const ColorSelect = ({
                         id,
                         value,
                         format,
                         onChange,
                         fromUserData,
                         enableAlpha = false,
                         placement = "bottomRight",
                         children
                     }) => {
    const handleChange = useCallback(
        data => onChange(generateChangeEvent(id, data.color, FIELD_TYPE.COLOR, fromUserData, format)),
        [id, fromUserData, format, onChange]
    );

    return (
        <>
            <ColorPicker
                enableAlpha={enableAlpha}
                color={value}
                onChange={handleChange}
                mode="RGB"
                placement={placement}
            />
            {children}
        </>
    );
};

ColorSelect.propTypes = {
    id: string.isRequired,
    value: string.isRequired,
    onChange: func.isRequired,
    enableAlpha: bool,
    placement: string
};

export default memo(ColorSelect);
