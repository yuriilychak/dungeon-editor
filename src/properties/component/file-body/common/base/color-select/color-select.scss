@import "../../../../../../sizes";
@import "../../../../../../colors";

.rc-color-picker-wrap {
  height: 20px;
  margin-right: $indent_l;
}

.rc-color-picker-trigger {
  padding: $indent_none;
  box-shadow: 0 0 0 2px $back-deep-dark inset;
  border: 1px solid $back-main;
  border-radius: $indent_none;
  width: 20px;
  height: 20px;
  box-sizing: border-box;
}

.rc-color-picker-panel-inner {
  border-radius: $indent_none;
  background-color: $back-main;
  border-color: $back-main;
  box-shadow: $indent_m $indent_m $indent_none $indent_none $back-deep-dark;
}

.rc-color-picker-panel-ribbon {
  border-radius: 0;
  border: 1px solid $back-light;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  color: $text-contrast;
  box-shadow: 0.5px 0 0 3px $back-deep-dark inset;
}

.rc-color-picker-panel-params input {
  border-radius: 0;
  border: 1px solid $back-light;
  background-color: $back-dark;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  color: $text-contrast;
}

.rc-color-picker-panel-params {
  border-radius: 0;
}

.rc-color-picker-panel-params-lable-hex,
.rc-color-picker-panel-params-lable-number {
  color: $text-contrast;
}

.rc-color-picker-panel-params-lable-number:hover {
  background-color: $back-light;
  border-radius: 0;
  border: none;
  box-shadow: none;
}

.rc-color-picker-panel-params input[type=number] {
  background-color: $back-dark;
  border-radius: 0;
  border: 1px solid $back-light;
  color: $text-contrast;
}

.rc-color-picker-panel-params-lable {
  color: $text-contrast;
}

.rc-color-picker-panel-board-handler {
  border-radius: 0;
  box-shadow: 0 0 0 3px black inset;
  border: 1px solid $back-light;
}

.rc-color-picker-panel-preview {
  border: none;
  border-radius: $indent_none;
}

.rc-color-picker-panel-preview span,
.rc-color-picker-panel-preview input[type=color] {
  box-shadow: 0 0 0 3px black inset;
  width: 30px;
  height: 30px;
  border: 1px solid $back-light;
  border-radius: 0;
}
