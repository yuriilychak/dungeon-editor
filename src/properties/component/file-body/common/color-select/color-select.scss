@import "../../../../../sizes";
@import "../../../../../colors";
@import "../../../../../utils";
@import "../../../../../other";

.properties-color-select-tooltip {
  position: absolute;
  top: calcSize(-10px);
  left: $half-size;
  transform: translateX(-$half-size);
  z-index: 2;
  background-color: $primary-main;
  font-size: $font-size_small;
  padding: $indent_m;
  opacity: 0;
  pointer-events: none;
}

.properties-color-select {
  position: relative;
  height: $item-size_small;

  &:hover .properties-color-select-tooltip {
    transition: opacity 0.3s, visibility 0.3s;
    opacity: 1;
  }
}

.rc-color-picker-wrap {
  height: $item-size_small;
}

.rc-color-picker-trigger {
  padding: $indent_none;
  box-shadow: 0 0 0 2px $back-deep-dark inset;
  border: $border-main;
  border-radius: $indent_none;
  width: $item-size_small;
  height: $item-size_small;
  box-sizing: border-box;
}

.rc-color-picker-panel-inner {
  border-radius: $indent_none;
  background-color: $back-main;
  border-color: $back-main;
  box-shadow: $indent_m $indent_m $indent_none $indent_none $back-deep-dark;
}

.rc-color-picker-panel-ribbon {
  border-radius: 0;
  border: $border-light;
  font-family: $font-family;
  color: $text-contrast;
  box-shadow: 0.5px 0 0 3px $back-deep-dark inset;
}

.rc-color-picker-panel-params input {
  border-radius: 0;
  border: $border-light;
  background-color: $back-dark;
  font-family: $font-family;
  color: $text-contrast;
}

.rc-color-picker-panel-params {
  border-radius: 0;
}

.rc-color-picker-panel-params-lable-hex,
.rc-color-picker-panel-params-lable-number {
  color: $text-contrast;
}

.rc-color-picker-panel-params-lable-number:hover {
  background-color: $back-light;
  border-radius: 0;
  border: none;
  box-shadow: none;
}

.rc-color-picker-panel-params input[type=number] {
  background-color: $back-dark;
  border-radius: 0;
  border: $border-light;
  color: $text-contrast;
}

.rc-color-picker-panel-params-lable {
  color: $text-contrast;
}

.rc-color-picker-panel-board-handler {
  border-radius: 0;
  box-shadow: 0 0 0 3px black inset;
  border: $border-light;
}

.rc-color-picker-panel-preview {
  border: none;
  border-radius: $indent_none;
}

.rc-color-picker-panel-preview span,
.rc-color-picker-panel-preview input[type=color] {
  box-shadow: 0 0 0 3px black inset;
  width: $item-size_medium - $indent_xs;
  height: $item-size_medium - $indent_xs;
  border: $border-light;
  border-radius: 0;
}
