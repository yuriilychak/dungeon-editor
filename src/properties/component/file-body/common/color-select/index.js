export { default as ColorSelect } from "./color-select";
