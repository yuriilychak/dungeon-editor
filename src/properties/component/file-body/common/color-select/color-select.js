import React, { memo } from "react";
import ColorPicker from "rc-color-picker";
import { string, func, bool } from "prop-types";

import {useChange} from "../../../../../hooks";

import "rc-color-picker/assets/index.css";
import "./color-select.scss";

export const ColorSelect = ({
    id,
    value,
    label,
    onChange,
    enableAlpha = false,
    placement = "bottomRight"
}) => {
    const handleChange = useChange(onChange, ({ color }) => color);

    return (
        <div className="properties-color-select">
            <ColorPicker
                enableAlpha={enableAlpha}
                color={value}
                onChange={handleChange}
                mode="RGB"
                placement={placement}
            />
            {label && (
                <span className="properties-color-select-tooltip">
                    {label}
                </span>
            )}
        </div>
    );
};

ColorSelect.propTypes = {
    id: string.isRequired,
    value: string.isRequired,
    onChange: func.isRequired,
    label: string,
    enableAlpha: bool,
    placement: string
};

export default memo(ColorSelect);
