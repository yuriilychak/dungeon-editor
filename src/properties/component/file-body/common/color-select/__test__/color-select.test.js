import { createMount } from "@material-ui/core/test-utils";
import React from "react";

import { ColorSelect } from "../color-select";

jest.mock("rc-color-picker", () => ({ onChange }) => (
    <button
        onClick={() => onChange({ color: "test color 2" })}
    >
        Mock Colorpicker
    </button>
));

describe("GIVEN ColorSelect from property section", () => {
    const mockChange = jest.fn();
    const wrapper = createMount()(
        <ColorSelect
            id="test"
            value="test color 1"
            label="test"
            onChange={mockChange}
        />
    );

    it("WHEN render component with default props THEN component should match snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("WHEN dispatch onChange THAN result should equal expected object", () => {
        wrapper.find("button").simulate("click");
        expect(mockChange).toHaveBeenCalledWith("test color 2");
    });
});
