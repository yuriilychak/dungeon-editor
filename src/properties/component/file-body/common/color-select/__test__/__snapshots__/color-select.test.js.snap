// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN ColorSelect from property section WHEN render component with default props THEN component should match snapshot 1`] = `"<div class=\\"properties-color-select\\"><button>Mock Colorpicker</button><span class=\\"properties-color-select-tooltip\\">test</span></div>"`;
